File: kreartif-api/controllers/karyaController.js
const { Karya } = require('../models/karyaModel');
const { Notifikasi } = require('../models/notifikasiModel');
const multer = require('multer');
const path = require('path');

// simpan file yang diupload
const storage = multer.diskStorage({
  destination: './uploads/',
  filename: (req, file, cb) => {
    cb(null, Date.now() + '-' + file.originalname);
  },
});

const upload = multer({ storage });
exports.uploadMiddleware = upload.single('file'); 

// GET semua karya
exports.getAllKarya = async (req, res) => {
  try {
    const karya = await Karya.findAll();
    res.json(karya);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
};

// GET karya milik user
exports.getMyKarya = async (req, res) => {
  try {
    const karya = await Karya.findAll({ where: { user_id: req.user.id } });
    res.json(karya);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
};

// GET karya berdasarkan ID
exports.getKaryaById = async (req, res) => {
  try {
    const { id } = req.params;
    const karya = await Karya.findByPk(id);
    if (!karya) return res.status(404).json({ msg: 'Karya tidak ditemukan' });
    res.json(karya);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
};

// POST: Buat karya baru
exports.createKarya = async (req, res) => {
  try {
    const { judul, deskripsi, kategori, file_url } = req.body;

    let finalFileUrl = file_url;

    if (req.file) {
      finalFileUrl = `${req.protocol}://${req.get('host')}/uploads/${req.file.filename}`;
    }

    if (!finalFileUrl) {
      return res.status(400).json({ error: 'File gambar atau URL harus disediakan.' });
    }

    const karya = await Karya.create({
      judul,
      deskripsi,
      kategori,
      file_url: finalFileUrl,
      user_id: req.user.id,
    });

    res.status(201).json(karya);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
};

// POST (khusus update): Update karya (dengan FormData)
exports.updateKarya = async (req, res) => {
  try {
    const { id } = req.params;
    const karya = await Karya.findByPk(id);

    if (!karya) return res.status(404).json({ msg: 'Karya tidak ditemukan' });
    if (karya.user_id !== req.user.id && req.user.role !== 'admin')
      return res.status(403).json({ msg: 'Tidak diizinkan' });

    const { judul, deskripsi, kategori } = req.body;

    const updateData = {
      judul,
      deskripsi,
      kategori,
    };

    // Hanya update file_url jika ada file baru
    if (req.file) {
      updateData.file_url = `${req.protocol}://${req.get('host')}/uploads/${req.file.filename}`;
    }

    await karya.update(updateData);
    res.json({ msg: 'Karya berhasil diperbarui', karya });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
};

// DELETE: Hapus karya dan notifikasi terkait
exports.deleteKarya = async (req, res) => {
  try {
    const { id } = req.params;
    const karya = await Karya.findByPk(id);

    if (!karya) return res.status(404).json({ msg: 'Karya tidak ditemukan' });
    if (karya.user_id !== req.user.id && req.user.role !== 'admin')
      return res.status(403).json({ msg: 'Tidak diizinkan' });

    // Hapus notifikasi terkait
    await Notifikasi.destroy({ where: { karya_id: id } });

    // Hapus karya
    await karya.destroy();

    res.json({ msg: 'Karya berhasil dihapus' });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
};

File: kreartif-api/controllers/karyaController.py
from fastapi import HTTPException, UploadFile, Request
from sqlalchemy.orm import Session, joinedload
from models.karyaModel import Karya
from models.notifikasiModel import Notifikasi
from uuid import uuid4
from typing import Optional
import os

UPLOAD_DIR = "./uploads"
os.makedirs(UPLOAD_DIR, exist_ok=True)


def get_all_karya(db: Session):
    karya_list = db.query(Karya).options(joinedload(Karya.user)).all()

    result = []
    for karya in karya_list:
        result.append({
            "id": karya.id,
            "judul": karya.judul,
            "deskripsi": karya.deskripsi,
            "kategori": karya.kategori,
            "file_url": karya.file_url,
            "user_id": karya.user_id,
            "username": karya.user.username if karya.user else None
        })
    return result


def get_my_karya(db: Session, current_user: dict):
    return db.query(Karya).filter(Karya.user_id == current_user["id"]).all()


def get_karya_by_id(db: Session, id: int):
    karya = db.query(Karya).filter(Karya.id == id).first()
    if not karya:
        raise HTTPException(status_code=404, detail="Karya tidak ditemukan")
    return karya


def create_karya(
    db: Session,
    current_user: dict,
    judul: str,
    deskripsi: str,
    kategori: str,
    file: Optional[UploadFile],
    request: Request,
    file_url: Optional[str] = None,
):
    final_file_url = file_url

    if file:
        filename = f"{uuid4().hex}_{file.filename}"
        filepath = os.path.join(UPLOAD_DIR, filename)
        with open(filepath, "wb") as f:
            f.write(file.file.read())
        final_file_url = f"{request.base_url}uploads/{filename}"

    if not final_file_url:
        raise HTTPException(status_code=400, detail="File atau URL harus disediakan")

    karya = Karya(
        judul=judul,
        deskripsi=deskripsi,
        kategori=kategori,
        file_url=final_file_url,
        user_id=current_user["id"],
    )

    db.add(karya)
    db.commit()
    db.refresh(karya)
    return karya


def update_karya(
    db: Session,
    current_user: dict,
    id: int,
    judul: Optional[str],
    deskripsi: Optional[str],
    kategori: Optional[str],
    file: Optional[UploadFile],
    request: Request,
):
    karya = db.query(Karya).filter(Karya.id == id).first()

    if not karya:
        raise HTTPException(status_code=404, detail="Karya tidak ditemukan")
    if karya.user_id != current_user["id"] and current_user["role"] != "admin":
        raise HTTPException(status_code=403, detail="Tidak diizinkan")

    if judul:
        karya.judul = judul
    if deskripsi:
        karya.deskripsi = deskripsi
    if kategori:
        karya.kategori = kategori

    if file:
        filename = f"{uuid4().hex}_{file.filename}"
        filepath = os.path.join(UPLOAD_DIR, filename)
        with open(filepath, "wb") as f:
            f.write(file.file.read())
        karya.file_url = f"{request.base_url}uploads/{filename}"

    db.commit()
    db.refresh(karya)
    return {"msg": "Karya berhasil diperbarui", "karya": karya}


def delete_karya(db: Session, current_user: dict, id: int):
    karya = db.query(Karya).filter(Karya.id == id).first()

    if not karya:
        raise HTTPException(status_code=404, detail="Karya tidak ditemukan")
    if karya.user_id != current_user["id"] and current_user["role"] != "admin":
        raise HTTPException(status_code=403, detail="Tidak diizinkan")

    db.query(Notifikasi).filter(Notifikasi.karya_id == id).delete()
    db.delete(karya)
    db.commit()
    return {"msg": "Karya berhasil dihapus"}
